### PR TITLE
Visibility widget

### PR DESCRIPTION
**Pull request by Hixie**
This attempts to address the confusion around how to hide a widget subtree.

Also, add alwaysIncludeSemantics to Opacity for consistency.

Also, create a SizedBox.shrink for this common case.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/basic.dart [hunk 1]
```diff
@@ -2252,8 +2287,10 @@ class SizedOverflowBox extends SingleChildRenderObjectWidget {
 ///
 /// See also:
 ///
-///  * The [catalog of layout widgets](https://flutter.io/widgets/layout/).
+///  * [Visibility], which can hide a child more efficiently (albeit less
+///    subtly).
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/basic.dart]: Why less subtly?

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+import 'package:flutter/foundation.dart';
+
+import 'basic.dart';
+import 'framework.dart';
+import 'ticker_provider.dart';
+
+/// Whether to show or hide a child.
+///
+/// By default, the [visible] property controls whether the [child] is included
+/// in the subtree or not; when it is not [visible], the [replacement] child
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/visibility.dart]: The replacement feature is not something I would expect to be done by a Visibility widget...
Do you expect that people will want to change maintainState and maintainSize for a Visibility widget dynamically? if not maybe it will be simpler to have the visibility widget only do show/hide and have a Replace widget for the replacement logic

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+  /// reach the [child] when it is hidden from the user. If this flag is set to
+  /// true, then touch events will nonetheless be passed through.
+  final bool maintainInteractivity;
+
+  @override
+  Widget build(BuildContext context) {
+    if (maintainSize) {
+      Widget result = child;
+      if (!maintainInteractivity) {
+        result = new IgnorePointer(
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/visibility.dart]: Should we be concerned about maintaining the tree structure when this is toggled.
Specifically if I'm not wrong toggling maintainInteractivity might destroy the state of child widgets without a key... I can imagine it being confusing to have a Visible widget with maintainState equals true destroy your state when you toggle interactivity...

File: packages/flutter/test/widgets/opacity_test.dart [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+            rect: Rect.fromLTRB(0.0, 0.0, 800.0, 600.0),
+            label: 'a',
+            textDirection: TextDirection.rtl,
+          )
+        ],
+      ),
+    ));
+    expect(find.byType(Opacity), paintsNothing);
+
+    // Opacity 0.0 without semantics: Nothing
```
Comment (amirh) [inline packages/flutter/test/widgets/opacity_test.dart]: nit: might be nice to separate all the different test cases into different testWidgets, with these comments being the title, I feel like it will be quicker to pinpoint the issue if one of these will fail.

File: packages/flutter/test/widgets/semantics_tester.dart [hunk 5]
```diff
@@ -600,6 +602,9 @@ class _HasSemantics extends Matcher {
     if (!doesMatch) {
       matchState['would-match'] = item.generateTestSemanticsExpressionForCurrentSemanticsTree(childOrder);
     }
+    if (item.tester.binding.pipelineOwner.semanticsOwner == null) {
```
Comment (amirh) [inline packages/flutter/test/widgets/semantics_tester.dart]: nit: these are the kind of changes that I feel will be better to be kept in a different PR, as it least for me it requires a major context switch from the current PR when reviewing (+ I'd imagine the commit message for this line will make more sense)

File: packages/flutter/test/widgets/visibility_test.dart [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    expect(find.text('a true', skipOffstage: false), findsOneWidget);
+    expect(find.byType(Visibility), paints..paragraph());
+    expect(tester.getSize(find.byType(Visibility)), const Size(800.0, 600.0));
+    expect(semantics, expectedSemanticsWhenPresent);
+    expect(log, <String>[]);
+    await tester.tap(find.byType(Visibility));
+    expect(log, <String>['tap']);
+    log.clear();
+
+    await tester.pumpWidget(new Visibility(child: testChild, visible: false));
```
Comment (amirh) [inline packages/flutter/test/widgets/visibility_test.dart]: nit: (as before) I'd love to have these as separate test cases with descriptive titles.

File: packages/flutter/test/widgets/visibility_test.dart [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    expect(find.byType(Placeholder), findsNothing);
+    expect(find.byType(Visibility), paints..paragraph());
+    expect(tester.getSize(find.byType(Visibility)), const Size(84.0, 14.0));
+    expect(semantics, expectedSemanticsWhenPresent);
+    expect(log, <String>[]);
+    await tester.tap(find.byType(Visibility));
+    expect(log, <String>['tap']);
+    log.clear();
+
+    await tester.pumpWidget(new Center(child: new Visibility(child: testChild, visible: false, maintainState: true, maintainAnimation: true, maintainSize: true, maintainInteractivity: true, maintainSemantics: true)));
```
Comment (amirh) [inline packages/flutter/test/widgets/visibility_test.dart]: We should test that state is maintained

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+///
+///  * [Opacity], which can stop its child from being painted.
+///  * [Offstage], which can stop its child from being laid out or painted.
+///  * [TickerMode], which can stop its child from being animated.
+///  * [ExcludeSemantics], which can hide the child from accessibility tools.
+///  * [IgnorePointer], which can disable touch interactions with the child.
+///
+/// Using this widget is not necessary to hide children. The simplest way to
+/// hide a child is just to not include it, or, if a child _must_ be given (e.g.
+/// because this is a [StatelessWidget]) then to use [SizedBox.shrink] instead
```
Comment (xster) [inline packages/flutter/lib/src/widgets/visibility.dart]: the 'this' in this sentence is not clear anymore. I think you're now referring a new hypothetical parent instead of this widget specifically.

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+  ///
+  /// The normal behavior is to replace the widget with a zero by zero box
+  /// ([SizedBox.shrink]).
+  ///
+  /// See also:
+  ///
+  ///  * [AnimatedCrossFade], which can animate between two children.
+  final Widget replacement;
+
+  /// Switches between showing the [child] or hiding it.
```
Comment (xster) [inline packages/flutter/lib/src/widgets/visibility.dart]: Perhaps add another paragraph to explain that 'hiding' here is a nuanced word since in some configurations, the child is effectively removed rather than strictly 'hidden'

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+  /// not [visible].
+  ///
+  /// To set this, [maintainState] must also be set.
+  ///
+  /// Keeping animations active when the widget is not visible is even more
+  /// expensive than only maintaining the state.
+  ///
+  /// One example when this might be useful is if the subtree contains an
+  /// animated [Image] that must be kept synchronized with other animated images
+  /// elsewhere in the interface.
```
Comment (xster) [inline packages/flutter/lib/src/widgets/visibility.dart]: The goal isn't very clear on first read. 

Perhaps s/./ when that image becomes visible again./

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+  /// ([SizedBox.shrink]).
+  ///
+  /// See also:
+  ///
+  ///  * [AnimatedCrossFade], which can animate between two children.
+  final Widget replacement;
+
+  /// Switches between showing the [child] or hiding it.
+  ///
+  /// The `maintain` flags should be set to the same values regardless of the
```
Comment (xster) [inline packages/flutter/lib/src/widgets/visibility.dart]: This part seems really important. Perhaps copy paste into the class's dart doc too so people exploring what this widget does can know right away what it isn't for.

Also, is it worth creating a StatelessElement.update override to debug assert that these indeed don't change (or show a warning at least)?

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+  final bool maintainAnimation;
+
+  /// Whether to maintain space for where the widget would have been.
+  ///
+  /// To set this, [maintainAnimation] must also be set.
+  ///
+  /// Maintaining the size when the widget is not [visible] is roughly as
+  /// expensive as keeping animations running, and may in some circumstances be
+  /// cheaper if the subtree is simple and the [visible] property is frequently
+  /// toggled, since it avoids triggering a layout change when the [visible]
```
Comment (xster) [inline packages/flutter/lib/src/widgets/visibility.dart]: It's not clear after reading this sentence what's the choice the developer needs to make or that maintainSize implies maintainAnimation.

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+  /// property is toggled. However, if the [child] subtree is not trivial then
```
Comment (xster) [inline packages/flutter/lib/src/widgets/visibility.dart]: The sentence transition is a bit strange. The 'However' makes it seem like you're continuing the previous discussion of whether to maintain both size and animation or just maintain animation but in fact you're now talking about something different (which is don't maintain anything at all)

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 3 above)
Comment (amirh) [reply packages/flutter/lib/src/widgets/visibility.dart]: Ohh I just saw your comment saying that the maintain flags should not be changed.
I'm a bit concerned that people (like I just did) might overlook that part in the comments and get confused by this.
I wonder if we can come up with an API where this will be more clear you're doing something wrong by changing the flags...?
One wild idea (which is probably bad) would be to have a separate widget class per maintain configuration....
Another option is maybe to make this a stateful widget just for the sake of asserting that the maintain flags does not change...?

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+///    subtree changes.
+///  * [AnimatedCrossFade], which can fade between two specific children.
+class Visibility extends StatelessWidget {
+  /// Control whether the given [child] is [visible].
+  ///
+  /// The [child] and [replacement] arguments must not be null.
+  ///
+  /// The boolean arguments must not be null.
+  ///
+  /// The [maintainSemantics] and [maintainInteractivity] arguments can only be
```
Comment (xster) [inline packages/flutter/lib/src/widgets/visibility.dart]: I would document the reverse dependencies in the fields' docs too. 

e.g. in maintainState: it needs to be true in order to maintain size

etc

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 3 above)
Comment (amirh) [reply packages/flutter/lib/src/widgets/visibility.dart]: Another potential alternative: if maintainState is true wrap child with a keyed widget and allow to modify the maintain flags

File: packages/flutter/lib/src/widgets/basic.dart [hunk 15]
```diff
@@ -1711,6 +1730,12 @@ class SizedBox extends SingleChildRenderObjectWidget {
       height = double.infinity,
       super(key: key, child: child);
 
+  /// Creates a box that will become as small as its parent allows.
+  const SizedBox.shrink({ Key key, Widget child })
```
[the remark this answers is not in the document]
Comment (slightfoot) [reply packages/flutter/lib/src/widgets/basic.dart]: @xster We've been using `Empty` and `Blank` as widgets in our app's which are just inherited from `SizedBox` with no width and height. Those could be quite good names as a synonym for a "NullWidget".

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 16]
```diff
@@ -756,13 +763,26 @@ class RenderOpacity extends RenderProxyBox {
...
+      return;
+    _alwaysIncludeSemantics = value;
+    markNeedsSemanticsUpdate();
+  }
 
   @override
   void paint(PaintingContext context, Offset offset) {
     if (child != null) {
-      if (_alpha == 0)
+      if (_alpha == 0) {
```
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/lib/src/rendering/proxy_box.dart]: The style guide these days says do whatever you want on this issue, because people had Strong Opinions on both side and it didn't seem worth fighting over. In this case I added the brace so that the method was internally consistent which I feel aids with readability (the other case has two statements so can't lose its braces).

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 8 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/visibility.dart]: Good call. Fixed.

File: packages/flutter/lib/src/widgets/basic.dart (hunk 1 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/basic.dart]: Added: "because it is either visible or hidden, rather than allowing fractional opacity values"

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 2 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/visibility.dart]: I wouldn't expect anything but `visible` to be changed dynamically.
`replacement` is there because it's not clear to me that `SizedBox.shrink` is always the right choice for the replacement (in the case of `maintainState` being false).

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 3 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/visibility.dart]: What would the use case be for changing the flags?

File: packages/flutter/test/widgets/opacity_test.dart (hunk 4 above)
Comment (Hixie) [reply packages/flutter/test/widgets/opacity_test.dart]: I specifically wanted them in the same testWidgets because one of the things being tested is whether you can actually change the values dynamically.

File: packages/flutter/test/widgets/semantics_tester.dart (hunk 5 above)
Comment (Hixie) [reply packages/flutter/test/widgets/semantics_tester.dart]: Ah, interesting. Ok. I included them because they were changes I needed to make to write the tests. Happy to split them out if you'd prefer.

File: packages/flutter/test/widgets/visibility_test.dart (hunk 7 above)
Comment (Hixie) [reply packages/flutter/test/widgets/visibility_test.dart]: will add

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 10 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/visibility.dart]: turns out this whole paragraph was dumb because images always stay in sync anyway. I've rewritten it.

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 11 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/visibility.dart]: Done and done.

I don't see why we'd assert that they can't change. It's a pretty common pattern to have flags like this which, if changed, will cause the subtree to be discarded. At the most trivial, `key` is one such property.

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 12 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/visibility.dart]: reworded to make this clearer.

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 13 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/visibility.dart]: removed "However"

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 14 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/visibility.dart]: done

File: packages/flutter/lib/src/widgets/basic.dart (hunk 15 above)
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/lib/src/widgets/basic.dart]: It's generally even better to just not include any widget at all.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 17]
```diff
@@ -146,15 +146,14 @@ class Directionality extends InheritedWidget {
...
-/// [Opacity] animations should be built using [AnimatedOpacity] rather than
-/// manually rebuilding the [Opacity] widget.
-///
 /// Animating an [Opacity] widget directly causes the widget (and possibly its
 /// subtree) to rebuild each frame, which is not very efficient. Consider using
 /// an [AnimatedOpacity] instead.
 ///
 /// See also:
 ///
+///  * [Visibility], which can hide a child more efficiently (albeit less
```
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/lib/src/widgets/basic.dart]: clarified

**Comment by Hixie** [comment]
There's no bug. This is inspired by various recent discussions.

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 3 above)
Comment (amirh) [reply packages/flutter/lib/src/widgets/visibility.dart]: Just the general approach that if it's there it will be used...

**Review by amirh — COMMENTED** [review]
LGTM
I'd consider a rename maintainState to something that sounds like maintainStateAcrossVisibilityChanges but I don't have a good short name :smile:

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 11 above)
Comment (xster) [reply packages/flutter/lib/src/widgets/visibility.dart]: I think the difference in this case is that it's a side effect in other cases but is a breakage in the semantically stated goal (of maintainingState) in this case. It would be debug-helpful to not have it fail silently.

Also, not so much a feedback but just curious for setting precedents, do we have an existing pattern for cases where changing a parameter on ephemeral widgets between frames causes misbehaviors in that widget's stated goal?

As a developer, I think I'd have that preference for that `key` example too. If I had an implicitly animating widget and I change the key and broke the animation, it would be nice to at least get a message at debug time if not an assert to say that changing key on the same widget type at the same spot in the tree is really weird, are you sure you want to do it?

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+  /// doing so will result in a subtree shape change).
+  ///
+  /// Unless [maintainState] is set, the [child] subtree will be disposed
+  /// (removed from the tree) while hidden.
+  final bool visible;
+
+  /// Whether to maintain the [State] objects of the [child] subtree when it is
+  /// not [visible].
+  ///
+  /// Keeping the state of the subtree is expensive. It should only be
```
Comment (xster) [inline packages/flutter/lib/src/widgets/visibility.dart]: Would be a nice place to say why it's expensive if the reader would think at this point that it's not rebuilding most of the time, it's not laying out and it's not painting. 

Also, I'm thinking how do we balance the relative weight of the word 'expensive' we use around the codebase (for instance relative to `This effect is relatively expensive` on BackdropFilter which out of thin air I'm guesstimating that it's more expensive than this)

File: packages/flutter/lib/src/widgets/visibility.dart [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+  /// subtree to be lost (and a new instance of the subtree, with new [State]
+  /// objects, to be immediately created if [visible] is true).
+  final bool maintainState;
+
+  /// Whether to maintain animations within the [child] subtree when it is
+  /// not [visible].
+  ///
+  /// To set this, [maintainState] must also be set.
+  ///
+  /// Keeping animations active when the widget is not visible is even more
```
Comment (xster) [inline packages/flutter/lib/src/widgets/visibility.dart]: These relative qualifiers are great

File: packages/flutter/lib/src/widgets/basic.dart (hunk 15 above)
[the remark this answers is not in the document]
Comment (slightfoot) [reply packages/flutter/lib/src/widgets/basic.dart]: I agree. I think @xster was talking about when you have to pass a child widget. Quite a lot of the time child is not optional and passing null will not work. This is the case for this "NullWiget"

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 11 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/visibility.dart]: Changing the key isn't really weird, it's literally how you indicate that you want to throw away the state.

Another example would be the widgets you build in `MaterialApp.builder`, if you change some of the flags on `MaterialApp` that change the tree shape. Or pretty much any of the flags on `Container`.

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 18 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/visibility.dart]: i've tried to clarify this further

File: packages/flutter/lib/src/widgets/visibility.dart (hunk 11 above)
Comment (xster) [reply packages/flutter/lib/src/widgets/visibility.dart]: `<feel free to ignore and submit. This is more educational for me than anything>`
* For the MaterialApp and Container etc, there's no conflict with the user's intent. There are side effects but you still correctly get a 'Container' when you change how the container draws itself and affects its child. Here, we're potentially breaking the stated intent of the user (of maintaining state). 

* I'm not sure changing key = losing state is super obvious for the layperson. We describe how to preserve state via keys in https://docs.flutter.io/flutter/widgets/StatefulWidget-class.html but maybe we should point it out more explicitly that changing key implies losing state and/or describe in https://docs.flutter.io/flutter/widgets/Widget/key.html that removing elements from the tree also implies losing state. 
`</>`
